tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -0,0 +1,610 @@
+    op_to_node_id.try_emplace(&op, op_list.size());
+    op_list.push_back(&op);
+  }
+  cycle_detector.reset(new GraphCycles(op_list.size()));
+  for (int64_t node_id = 0; node_id < op_list.size(); node_id++) {
+    Operation* op = op_list[node_id];
+    for (Value operand : GetAllPossibleUsedValues(op)) {
+      Operation* operand_op = operand.getDefiningOp();
+      // Only consider the operand_op inside the target block.
+      auto iter = op_to_node_id.find(operand_op);
+      if (iter == op_to_node_id.end()) {
+        continue;
+      }
+      cycle_detector->InsertEdge(iter->second, node_id);
+    }
+  }
+}
+
+struct DotCluster {
+  DotCluster(Operation* op, int op_id) : leader_op_id(op_id) {
+    ops.push_back(op);
+  }
+
+  // Merges `other` into this cluster, and clears `other`.
+  void merge(DotCluster& other) {
+    ops.insert(ops.end(), other.ops.begin(), other.ops.end());
+    other.ops.clear();
+  }
+
+  // ID of the representative node of this cluster.
+  int leader_op_id;
+
+  // Dot ops to be batched.
+  SmallVector<Operation*> ops;
+};
+
+bool TryMergeDotClusters(DotCluster& dst, DotCluster& src,
+                         std::unique_ptr<GraphCycles>& cycle_detector) {
+  // Check cycle.
+  int64_t dst_id = dst.leader_op_id;
+  int64_t src_id = src.leader_op_id;
+  auto optional_merged_id = TryMergeNode(cycle_detector.get(), dst_id, src_id);
+  if (!optional_merged_id.hasValue()) {
+    // It forms a cycle.
+    return false;
+  }
+  dst.merge(src);
+  dst.leader_op_id = *optional_merged_id;
+  return true;
+}
+
+// The `MergingShapeEqualityMap` is a map whose value is a list of dots with
+// same `MergingShape`.
+template <typename MergingShapeEqualityMap>
+bool BuildDotClusters(Block* block,
+                      const MergingShapeEqualityMap& equal_merge_shape_map,
+                      SmallVector<DotCluster>& merging_clusters) {
+  merging_clusters.clear();
+  // Dot ops in `equal_merge_shape_map` can be merged together if the merging
+  // does not introduce cycle.
+
+  // Form cycle detector.
+  std::unique_ptr<GraphCycles> cycle_detector(new GraphCycles(0));
+  DenseMap<Operation*, int64_t> op_to_node_id;
+  BuildBlockGraphCycles(block, cycle_detector, op_to_node_id);
+
+  // Find merge clusters.
+  for (auto& dots_can_merge : equal_merge_shape_map) {
+    auto ops = dots_can_merge.second;
+    if (ops.size() < 2) {
+      continue;
+    }
+    SmallVector<DotCluster> clusters;
+    for (auto op : ops) {
+      clusters.emplace_back(op.getOperation(),
+                            op_to_node_id[op.getOperation()]);
+    }
+    for (int64_t i = 0; i < clusters.size(); i++) {
+      auto& merged = clusters[i];
+      if (merged.ops.empty()) {
+        continue;
+      }
+      for (int64_t j = i + 1; j < clusters.size(); j++) {
+        auto& to_merge = clusters[j];
+        if (to_merge.ops.empty()) {
+          continue;
+        }
+        // Try merge.
+        TryMergeDotClusters(merged, to_merge, cycle_detector);
```

Comment:
what will happen if TryMergeDotClusters() returns false? 
is it expected for BuildDotClusters() to always return true?

Reply:
OK, i see the return value for both TryMergeDotClusters() and BuildDotClusters() are not used. 
better set to void to avoid redundant warnings.